README.md

```diff
@@ -1,84 +1,155 @@
 
 ```
-<header>
-<arbitrary-stream-data>
+> <multicodec-header-of-multistream>
+> <multicodec-header-for-whatever-protocol-that-we-want-to-speak>
 ```
 
-### The header
+for example:
 
-The header has three parts:
-
-- `hdr-len` - a varint length, in bytes, for security and binary protocols.
-- `path` - the path of the protocol in a universal namespace. UTF-8. must start with a slash.
-- `\n` - a newline at the end, for the benefit of text protocols **(included in hdr-len)**.
-
-It looks like this:
 ```
-<hdr-len><path>\n
+> /ipfs/QmdRKVhvzyATs3L6dosSb6w8hKuqfZK2SyPVqcYJ5VLYa2/multistream-select/0.3.0
+> /ipfs/QmVXZiejj3sXEmxuQxF2RjmFbEiE9w7T82xDn3uYNuhbFb/ipfs-dht/0.2.3
 ```
 
-So the full protocol is:
+- The `<multicodec-header-of-multistream>` ensures a protocol selection is happening.
+- The `<multistream-header-for-whatever-protocol-is-then-selected>` hopefully describes a valid protocol listed. Otherwise we return a `na`("not available") error:
 
 ```
-<hdr-len><path>\n
-<arbitrary-stream-data>
+na\n
+
+# in hex (note the varint prefix = 3)
+# 0x036e610a
 ```
 
-For example:
+for example:
 
 ```sh
-# this header:
+# open connection + send multicodec headers, inc for a protocol not available
+> /ipfs/QmdRKVhvzyATs3L6dosSb6w8hKuqfZK2SyPVqcYJ5VLYa2/multistream-select/0.3.0
+> /ipfs/QmVXZiejj3sXEmxuQxF2RjmFbEiE9w7T82xDn3uYNuhbFb/some-protocol-that-is-not-available
+
+# open connection + signal protocol not available.
+< /ipfs/QmdRKVhvzyATs3L6dosSb6w8hKuqfZK2SyPVqcYJ5VLYa2/multistream-select/0.3.0
+< na
+
+# send a selection of a valid protocol + upgrade the conn and send traffic
+> /ipfs/QmVXZiejj3sXEmxuQxF2RjmFbEiE9w7T82xDn3uYNuhbFb/ipfs-dht/0.2.3
+> <dht-traffic>
+> ...
+
+# receive a selection of the protocol + sent traffic
+< /ipfs/QmVXZiejj3sXEmxuQxF2RjmFbEiE9w7T82xDn3uYNuhbFb/ipfs-dht/0.2.3
+< <dht-traffic>
+< ...
+```
 
-/echo/1.0
+#### Listing
 
-# with hexdump -C style inspection:
-0a 2f 65 63 68 6f 2f 31  2e 30 0a                 |./echo/1.0.|
+It is also possible to "list" the available protocols. A list message is simply:
 
-0a    - varint hdr len   10 bytes
-2f-30 - "/echo/1.0"       9 bytes
-0a    - newline           1 byte
 ```
+ls\n
 
-### The protocol path
+# in hex (note the varint prefix = 3)
+0x036c730a
+```
 
-`multistream` allows us to specify different protocols in a universal namespace, that way being able to recognize, multiplex, and embed them easily. We use the notion of a `path` instead of an `id` because it is meant to be a Unix-friendly URI.
+So a remote side asking for a protocol listing would look like this:
 
-A good path name should be decipherable -- meaning that if some machine or developer -- who has no idea about your protocol -- encounters the path string, they should be able to look it up and resolve how to use it.
+```sh
+# request
+<multistream-header-for-multistream-select>
+ls\n
+
+# response
+# TODO: maybe include a varint number of protocols here ?
```

Comment:
Protocol additions:
- respect both ls response constraints:
  - if the total size of ls response is larger than the total buffer, assume malformed. discard rest of response, only use LAST complete protocol listed within buffer size (NEVER use a partial protocol string)
  - if the total number of protocols in ls response is larger than number of protocols listed, assume malformed. discard rest of response. only use LAST complete protocol listed within the number of protocols.